sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridConsumer.cs

```diff
@@ -0,0 +1,301 @@
+                                }
+                                else
+                                {
+                                    cloudEventData = _objectSerializer.Deserialize(dataStream, typeOfEventData, cancellationToken);
+                                }
+                            }
+                        }
+                        // If a custom mapping doesn't exist, let's attempt to find the mapping for the deserialization function in the system event type mapping.
+                        else if (SystemEventTypeMappings.SystemEventDeserializers.TryGetValue(cloudEventInternal.Type, out Func<JsonElement, object> systemDeserializationFunction))
+                        {
+                            cloudEventData = systemDeserializationFunction(dataElement.Value);
+                        }
+                        // If no custom mapping was added, either return a primitive/string, or an object wrapped as BinaryData
+                        else
+                        {
+                            // If event data is not a primitive/string, return as BinaryData
+                            if (!TryGetPrimitiveFromJsonElement(dataElement.Value, out cloudEventData))
+                            {
+                                cloudEventData = BinaryData.FromStream(dataStream);
+                            }
+                        }
+                    }
+                    else if (dataElement?.ValueKind == JsonValueKind.Null) // Event has null data
+                    {
+                        cloudEventData = null;
+                        cloudEventInternal.Type = "";
+                    }
+                    else // Event has no data
+                    {
+                        cloudEventData = "";
+                        cloudEventInternal.Type = "";
+                    }
+                }
+
+                cloudEvents.Add(new CloudEvent(
+                    cloudEventInternal.Source,
+                    cloudEventInternal.Type)
+                {
+                    Id = cloudEventInternal.Id,
+                    Data = cloudEventData,
+                    Time = cloudEventInternal.Time,
+                    DataSchema = cloudEventInternal.Dataschema,
+                    DataContentType = cloudEventInternal.Datacontenttype,
+                    Subject = cloudEventInternal.Subject
+                });
+            }
+
+            return cloudEvents.ToArray();
+        }
+
+        private static bool TryGetPrimitiveFromJsonElement(JsonElement jsonElement, out object elementValue)
+        {
+            elementValue = null;
+            if (jsonElement.ValueKind == JsonValueKind.True || jsonElement.ValueKind == JsonValueKind.False)
+            {
+                elementValue = jsonElement.GetBoolean();
+            }
+            else if (jsonElement.ValueKind == JsonValueKind.Number)
+            {
+                var oki = jsonElement.TryGetInt32(out var vali);
```

Comment:
nit: why do we need these `oki` vars? can we just do `if (jsonElement.TryGetInt32(out var vali)'?